src/llvm-multiversioning.cpp

```diff
@@ -707,8 +748,15 @@ void CloneCtx::fix_gv_uses()
             auto info = uses.get_info();
             // We only support absolute pointer relocation.
             assert(info.samebits);
-            // And only for non-constant global variable initializers
-            auto val = cast<GlobalVariable>(info.val);
+            GlobalVariable *val;
+            if (auto alias = dyn_cast<GlobalAlias>(info.val)) {
+                rewrite_alias(alias);
+                uint32_t id;
+                std::tie(id, val) = get_reloc_slot(orig_f);
```

Comment:
[As I said](https://github.com/JuliaLang/julia/pull/37530#discussion_r487687922), there should be no need to do anything here, just `continue` should be fine. You can leave the `auto val = cast<GlobalVariable>(info.val);` line unchanged after the `GlobalAlias` branch.
The reason I said you don't need return value is because you don't need them, not because you can get it again after returning nothing......

Also, to be padentic, you want to pass in `orig_f` to `rewrite_alias` and use that to do the lookup instead. I think global alias allows constant expression and we accept it as long as it's a simple cast. (asserted above).
Since we are emitting a function, we **could** accept more complicated expressions here but that's not necessary for now....

Reply:
But then the GV never gets initialized because `gv_relocs.emplace_back(addr, fid);` doesn't get hit. And having `rewrite_alias` add a use of `orig_f` seems like it would invalidate the `uses` iterator?

Reply:
> `gv_relocs.emplace_back(addr, fid);`

This should not be needed and that's whole point of using the global slot everyone else uses instead of creating a new one. Special initialization of global variables are only needed because they are different from the normal ones returned by `get_reloc_slot`.

>  having rewrite_alias add a use of orig_f

Nothing like that should be needed and in fact it is not valid to do so. The global variables are all NULL initialized. `orig_f` isn't used by anyone.

Reply:
Then how does the runtime know how to initialize that slot? `get_reloc_slot` doesn't register it.
With `rewrite_alias` using `get_reloc_slot` (i.e. not creating a custom one) and:

```patch
@@ -750,9 +750,8 @@ void CloneCtx::fix_gv_uses()
             assert(info.samebits);
             GlobalVariable *val;
             if (auto alias = dyn_cast<GlobalAlias>(info.val)) {
-                rewrite_alias(alias);
-                uint32_t id;
-                std::tie(id, val) = get_reloc_slot(orig_f);
+                rewrite_alias(alias, orig_f);
+                continue;
             }
```

```
julia> Float16(1.0)

Thread 1 "julia" hit Breakpoint 1, 0x00007fffecc0a740 in __truncdfhf2 () from /home/tim/Julia/julia/build/release/usr/lib/julia/sys.so
(gdb) disassemble
Dump of assembler code for function __truncdfhf2:
=> 0x00007fffecc0a740 <+0>:     jmpq   *0x51db222(%rip)        # 0x7ffff1de5968 <jlcapi___truncdfhf2_21535.reloc_slot>
End of assembler dump.
(gdb) ni
0x0000000000000000 in ?? ()
(gdb) c
Continuing.

Thread 1 "julia" received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
```

Reply:
Hmmm, it does register it in `const_relocs`.

Reply:
I seems that the issue could be this function not being in `fvars`? I'm not sure if that's even supported.

Reply:
> I seems that the issue could be this function not being in `fvars`?

Yes, that seems to be the case. I had a quick look but the way `jl_sysimg_fvars` is populated in `jl_create_native` and how that ought to work with `jl_compile_extern_c` is not obvious.

Reply:
If I put the C wrapper in `fvars` (https://github.com/JuliaLang/julia/pull/37530/commits/8064e00e1aa63d86277a662c6c42f14ddbbe6c71), multiversioning seems to work: the original function now gets a proper ID, I get an entry in `jl_dispatch_reloc_slots`, etc

```
@jl_dispatch_fvars_offsets = constant [860 x i32] [i32 trunc (i64 sub (i64 ptrtoint (void ()* @jlcapi_dummy_ccallable_2883.clone_1 to i64), i64 ptrtoint (void ()* @jlcapi_dummy_ccallable_2883 to i64)) to i32), ...]

@jl_sysimg_fvars_base = alias i64, bitcast (void ()* @jlcapi_dummy_ccallable_2883 to i64*)
```

At run-time however the slot is not initialized and I still get a nullptr SEGV.

Reply:
Did some more debugging, and it looks like the slots generated for the newly-created alias GVs are never listed in `jl_dispatch_fvars_idxs`, so no relocation ever happens (the run-time logic just skips past them). Other GVs are explicitly added to `gv_relocs`, which is then used to populate `shared_relocs` resulting in an entry in `jl_dispatch_fvars_idxs`. By reusing `get_reloc_slot`, which adds to `const_relocs`, that doesn't happen.
Could you explain the logic how non-GV slots are chosen to be listed in `jl_dispatch_fvars_idxs`, and/or how to adapt that for this case? Just push them to `gv_relocs` manually?

Reply:
OK, I missed the mutation of the `fvars` in `get_func_id` which is why I thought this isn't supposed to work but yes it should work. The point was to basically use the logic for instruction relocation rather than global variable one since this is what you are doing. The missing part is that you need to record that this function needs to be relocated for the target. This is done by pushing the id into the `relocs` set for the group and the target. Something like,

```c++
    for (auto &grp: groups) {
        grp.relocs.insert(id);
        for (auto &tgt: grp.clones) {
            tgt.relocs.insert(id);
        }
    }
```

within `rewrite_alias` after you got the `id`.

Reply:
That works for target 0 | CLONE_ALL, but doesn't cover https://github.com/JuliaLang/julia/blob/2898b93264a4dfed09a30251af00b54f37cb1e1d/src/llvm-multiversioning.cpp#L1053-L1057

I guess I want `tgt.vmap[base_f] = base_f` here, but `map_get` is designed to skip over such identical mappings.